pkg/nimble/scanlist/nimble_scanlist.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2019 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     ble_nimble_scanlist
+ * @{
+ *
+ * @file
+ * @brief       Implementation of a list for NimBLE scan results
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ *
+ * @}
+ */
+#include <limits.h>
+
+#include "xtimer.h"
+#include "net/bluetil/ad.h"
+
+#include "nimble_scanlist.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static nimble_scanlist_entry_t _mem[NIMBLE_SCANLIST_SIZE];
+
+static clist_node_t _pool;
+static clist_node_t _list;
+
+static int _finder(clist_node_t *node, void *arg)
+{
+    const ble_addr_t *addr = (const ble_addr_t *)arg;
+    nimble_scanlist_entry_t *e = (nimble_scanlist_entry_t *)node;
+    if (ble_addr_cmp(&e->addr, addr) == 0) {
+        return 1;
+    }
+    return 0;
+}
+
+static nimble_scanlist_entry_t *_find(const ble_addr_t *addr)
+{
+    return (nimble_scanlist_entry_t *)clist_foreach(&_list, _finder,
+                                                    (void *)addr);
+}
+
+void nimble_scanlist_init(void)
+{
+    for (unsigned i = 0; i < (sizeof(_mem) / sizeof(_mem[0])); i++) {
+        clist_rpush(&_pool, &_mem[i].node);
```

Comment:
Can you give a reasoning why you use a list to manage the pool of available nodes instead of just using an array where you search for empty entries by iterating over it? While I certainly like the elegance of this approach, it seem to me to be an unnecessary level of complexity that might introduce bugs when not properly tested (which since there are no unittests provided we can't).

Reply:
I see it on the contrary: using a well-tested external list library is less error prone than using a custom array construct.

So my reasons:
- lists are more elegant
- lists are more efficient (except for the `nimble_scanlist_get_by_pos()`)
- lists are well tested
- no code duplication

Reply:
True, that would be a possibility in general. But in this particular case this is not an issue, as items are only taken from the pool, and the only operation to put them back is by calling clear(), so not really a high chance of loosing an item anywhere.

Hope you can live with this for now. At some point, we might as well look into how we can add unittest to these submodules in a nice way - lets see if this submodules survives until then :-)

Reply:
Yepp, I can live with it if your sign your name in blood that you jump to fix it when a problem occurs ;-).